protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java

```diff
@@ -107,23 +110,38 @@ public T parse(InputStream stream) {
             }
           }
         }
+        CodedInputStream cis;
+        if (stream instanceof ReadableBuffer) {
```

Comment:
Because of the way CIS works, I think this is either a no-op or objectively better.  There are like 3 copies in the current code path, this cuts it down to be at most 1 or 2 based on message size.  

Before (large message): 
1.  Copy from network to direct byte buffer (netty)
2.  Copy from Composite Byte buffer to CIS's byte array (4k chunks)
3.  Copy from 4k byte chunks into CIS's single large byte array (aka "the buffer")
4.  Copy from "the buffer" reading out into a ByteString or byte array to read out the field from the proto.

There is a lot of bad stuff going on in CIS.  The above code path also uses the constructor that thinks data is mutable, so it always copies in step 4.

After:
1.  Copy from network to direct byte buffer (netty)
2.  Copy from Composite Byte Buffer to single Heap byte buffer.
3.  Pass that array to CIS via unsafe.  (and is now has  immutable set)
4.  Copy from "the buffer" reading out into a ByteString or byte array to read out the field from the proto.

The after case also opens up the possibility of using aliasing, which would avoid the copy in step 4 in both cases (but may have other side effects).

Since Netty is passing Direct byte buffers instead of heap, we always have to copy anyway, because CIS is centered around have a byte array.  The only way to avoid this copy here is if netty was using Heap byte buffers, and there was only one (a.k.a. a tiny message).  The same goes for OkHTTP.  

All of this is to say that I think copying here is almost certainly the right thing to do, and we would have to have a very contorted scenario for it to not be right.  @nmittler could maybe change the Protobuf API to make this less painful? :)